File: hotalg/wish/L240.py
"""
编写一个高效的算法来搜索 m x n 矩阵 matrix 中的一个目标值 target 。该矩阵具有以下特性：

每行的元素从左到右升序排列。
每列的元素从上到下升序排列。

"""
class Solution(object):
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        return self.search(matrix, 0, len(matrix) - 1, 0, len(matrix[0]) - 1, target)

    def search(self, matrix, sr, er, sc, ec, target):
        if sr > er or sc > ec:
            return False
        elif sr == er and sc == ec:
            return matrix[sr][sc] == target
        else:
            if matrix[sr][ec] == target:
                return True
            elif matrix[sr][ec] > target:
                return self.search(matrix, sr, er, sc, ec - 1, target)
            else:
                return self.search(matrix, sr + 1, er, sc, ec, target)

File: algorithm/src/main/java/linshen/binary/L240.java
package linshen.binary;

/*
编写一个高效的算法来搜索 m x n 矩阵 matrix 中的一个目标值 target 。该矩阵具有以下特性：

每行的元素从左到右升序排列。
每列的元素从上到下升序排列。
 */

public class L240 {
    public boolean searchMatrix(int[][] matrix, int target) {
        return find(matrix, 0, matrix[0].length-1, target);
    }

    private boolean find(int[][] matrix, int x, int y, int target) {
        if(x < 0 || x >= matrix.length || y < 0 || y >= matrix[0].length)return false;
        if(matrix[x][y] == target)return true;
        else if(matrix[x][y] > target) {
            return find(matrix, x, y-1, target);
        } else {
            return find(matrix, x + 1, y, target);
        }
    }
}
